django/utils/dateformat.py

```diff
@@ -30,11 +30,17 @@
 
 
 class Formatter(object):
+
+    def format_piece(self, piece):
+        if type(self.data) is datetime.date and hasattr(TimeFormat, piece):
+            raise TypeError('The format for date objects may not contain time-related format specifiers')
+        return piece
+
     def format(self, formatstr):
         pieces = []
         for i, piece in enumerate(re_formatchars.split(force_text(formatstr))):
             if i % 2:
-                pieces.append(force_text(getattr(self, piece)()))
+                pieces.append(force_text(getattr(self, self.format_piece(piece))()))
             elif piece:
                 pieces.append(re_escaped.sub(r'\1', piece))
```

Comment:
Do we need to check the `piece` in this branch too? If so, check it before the `if ` statement.